src/shared/unit-file.c

```diff
@@ -21,3 +22,60 @@ bool unit_type_may_template(UnitType type) {
                       UNIT_TIMER,
                       UNIT_PATH);
 }
+
+int unit_validate_alias_symlink_and_warn(const char *filename, const char *target) {
+        const char *src, *dst;
+        UnitType src_type, dst_type;
+        int r, r2;
+
+        /* Check if the *alias* symlink is valid. This applies to symlinks like
+         * /etc/systemd/system/dbus.service → dbus-broker.service, but not to .wants or .requires symlinks
+         * and such. Neither does this apply to symlinks which *link* units, i.e. symlinks to outside of the
+         * unit lookup path.
+         */
+
+        src = basename(filename);
+        dst = basename(target);
+
+        /* Reject cases where the alias or the target unit have instance names. */
+        r = unit_name_to_instance(src, NULL);
+        if (r < 0)
+                return log_notice_errno(r, "%s: not a valid unit name \"%s\": %m", filename, src);
+        if (r == UNIT_NAME_INSTANCE)
+                return log_notice_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "%s: unit symlink has instance name, rejecting.", filename);
+
+        src_type = unit_name_to_type(src);
+        assert(src_type >= 0); /* unit_name_to_instance() checked the suffix already */
+        if (!unit_type_may_template(src_type))
+                return log_notice_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "%s: templates not allowed for %s units, rejecting.",
+                                        filename, unit_type_to_string(src_type));
+
+        r2 = unit_name_to_instance(dst, NULL);
+        if (r2 < 0)
+                return log_notice_errno(r2 == -EINVAL ? SYNTHETIC_ERRNO(EXDEV) : r2,
+                                        "%s points to \"%s\" which is not a valid unit name: %m", filename, dst);
+        if (r2 == UNIT_NAME_INSTANCE)
+                return log_notice_errno(SYNTHETIC_ERRNO(EXDEV),
+                                        "%s: unit symlink target \"%s\" has instance name, rejecting.", filename, dst);
```

Comment:
hmm, this is a new limitation that I am not sure I agree with? i mean, we previously said as long as the instance name matches you can have aliases, and that when searching for a file we'd always look for the name with the instance first, and without it as fallback. Why take this away? Are you sure this is nowhere used?

Reply:
OK, I'll remove this limitation. (BTW, thanks, this is the kind of comments I was looking for).

Reply:
Oh, I remember why this is a problem. Normally aliases are functions (in the mathematical sense, i.e. there can by only one y for any given x). This is nicely enforced by the fact that a symlink can only point one way, and once we find the symlink, we don't look for any fragments with lower priority. But once we allow symlinks for aliases, we can have `a@foo.service → b@foo.service` and at the same time `a@.service → c@.service`, while both `b@.service` and `c@service` can have fragments on disk. So are now `a@.service`, `b@.service`, `c@.service` now all aliases? If yes, do we pick the fragment for `b@.service` or `c@.service` ?

Reply:
Actually, this can be even simpler: `a@.service` may exist on disk, and `b@foo.service` may be a symlink to `b@foo.service`. Now we have two candidate fragments again.

Maybe we should allow this, check that the instance alias matches the main alias, but warn?

Reply:
I'd say the rule is that if `a@foo.service` → `b@foo.service` exists that `a@.service` is never itself read (but the drop-ins for that are.)

i.e. in your example, I think we should read:

* `b@foo.service`
* `a@foo.service.d/*.conf`
* `b@foo.service.d/*.conf`

But not:

* `c@foo.service`
* 'c@foo.service.d/*`

i.e. we only collect the drop-ins on the symlink chain we actually follow, and we stop searching as soon as we found the first unit file that is not a symlink. that means we'd never be tempted to resolve `a@.service` at all, since we already found a unit file by traversing `a@foo.service`.

Reply:
Dunno, the fact that we read `a@.service.d/*.conf` but not `a@.service` seems very arbitrary. Before, it didn't matter if a setting was defined in the fragment or in the drop-in (apart from ordering issues). But now the dropins start living a life independent of the main unit.

Reply:
Hmm, I can't convince myself that this is OK.
We either consider the instanced unit `b@foo` independent of the template `a@.service`, or not.
If the first, then we shouldn't load `a@.service` or `a@.service.d`. If the second, we should load both.